lib/import-map.php

```diff
@@ -0,0 +1,133 @@
+	global $import_map_entries;
+
+	$import_map_entries[ $key ] = $value;
+}
+
+/**
+ * Injects the import map to admin pages.
+ */
+function gutenberg_inject_import_map() {
+	global $import_map_entries;
+
+	if ( empty( $import_map_entries ) ) {
+		return;
+	}
+
+	echo '
+<script type="importmap">
+{
+	"imports": ' . json_encode( $import_map_entries ) . '
+}
+</script>
+	';
+}
+
+add_action( 'admin_enqueue_scripts', 'gutenberg_inject_import_map' );
+
+gutenberg_register_import_map_entry( 'lodash', 'https://cdn.skypack.dev/lodash' );
+gutenberg_register_import_map_entry( 'react', 'https://cdn.skypack.dev/react@17.0.1' );
+gutenberg_register_import_map_entry( 'react-dom', 'https://cdn.skypack.dev/react-dom@17.0.1' );
+gutenberg_register_import_map_entry( 'moment', 'https://cdn.skypack.dev/moment/moment' );
+gutenberg_register_import_map_entry( 'moment-timezone/moment-timezone', 'https://cdn.skypack.dev/moment-timezone/moment-timezone' );
+gutenberg_register_import_map_entry( 'moment-timezone/moment-timezone-utils', 'https://cdn.skypack.dev/moment-timezone/moment-timezone-utils' );
+
+/**
+ * Convert string to in camel-case, useful for class name patterns.
+ *
+ * @param string $string Target string.
+ *
+ * @return string Camel-case string.
+ */
+function gutenberg_to_camel_case( $string ) {
+	$string = str_replace( '-', ' ', $string );
+	$string = str_replace( '_', ' ', $string );
+	$string = lcfirst( ucwords( strtolower( $string ) ) );
+	$string = str_replace( ' ', '', $string );
+	return $string;
+}
+
+foreach ( glob( gutenberg_dir_path() . 'build/esm/*.js' ) as $path ) {
+	$package_name = substr( basename( $path ), 0, -3 );
+	gutenberg_register_import_map_entry( '@wordpress/' . $package_name, plugins_url( 'build/esm', __DIR__ ) . '/' . $package_name . '.js' );
+}
+
+/**
+ * Hichjack the scripts initialization to use the esmodules instead.
+ *
+ * @param string $tag    The `<script>` tag for the enqueued script.
+ * @param string $handle The script's registered handle.
+ * @param string $src    The script's source URL.
+ *
+ * @return string Updated tag.
+ */
+function gutenberg_inject_module_scripts( $tag, $handle, $src ) {
+	$packages = array();
+	foreach ( glob( gutenberg_dir_path() . 'build/esm/*.js' ) as $path ) {
+		$package_name = substr( basename( $path ), 0, -3 );
+		$packages[]   = 'wp-' . $package_name;
+	}
+	$vendors                 = array( 'lodash', 'react', 'react-dom', 'moment', 'moment-timezone' );
+	$export_default_packages = array(
+		'api-fetch',
+		'deprecated',
+		'dom-ready',
+		'redux-routine',
+		'token-list',
+		'server-side-render',
+		'shortcode',
+		'warning',
+	);
+
+	if ( in_array( $handle, $packages, true ) ) {
+		$package_name = substr( $handle, 3 );
+		$import       = in_array( $package_name, $export_default_packages, true ) ? 'mod' : '* as mod';
+		// Replace all scripts with a module script to defer them.
+		$tag = str_replace( '<script ', '<script type="module" ', $tag );
```

Comment:
`script_loader_tag` is only called once and includes before, after and the script itself:

for the before/after I just mark the script as "type module" to defer it, but for the script source itself, I replace it with the module import.